File: problems/024/024.c
#include <stdio.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */

#define N_MAX 10

int main()
{
    const int i = 1E6;
    int numbers[N_MAX] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };

    long int MAX_PERMUTATION = 1;
    for (int j = 2; j <= N_MAX; j++)
        MAX_PERMUTATION *= j;

    int n = N_MAX;
    long int n_product = N_MAX;

    char digits[N_MAX+1];

    int index  = (int)( (i - 1) * n_product / MAX_PERMUTATION );
    digits[0]  = numbers[index] + '0';

    for (int j = 1; j < N_MAX; j++)
    {
        n--;
        n_product *= n;

        if (index < (N_MAX - 1))
            memmove( &numbers[index], 
                     &numbers[index+1], 
                     (N_MAX - index - 1) * sizeof(int));

        index = (int)( (i - 1) * n_product / MAX_PERMUTATION ) % n;

        digits[j] = numbers[index] + '0';
    }

    printf("%s\n", digits);

    return 0;
}

File: problems/024/024.f90
PROGRAM F024

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    INTEGER, PARAMETER :: N_MAX = 10
    INTEGER, PARAMETER :: I     = 1000000

    CHARACTER (LEN=1), DIMENSION(N_MAX) :: NUMBERS = &
    (/ "0", "1", "2", "3", "4", "5", "6", "7", "8", "9" /)

    INTEGER                    :: N               = N_MAX
    INTEGER (KIND=8)           :: N_PRODUCT       = N_MAX
    INTEGER (KIND=8)           :: MAX_PERMUTATION = 1

    CHARACTER (LEN=N_MAX)      :: DIGITS_VAL

    INTEGER :: J, INDEX_VAL

    DO J = 2, N_MAX
        MAX_PERMUTATION = MAX_PERMUTATION * J
    END DO

    INDEX_VAL = FLOOR(REAL((I - 1) * N_PRODUCT) / REAL(MAX_PERMUTATION)) + 1

    DIGITS_VAL(1:1) = NUMBERS(INDEX_VAL)

    DO J = 2, N_MAX

        N = N - 1
        N_PRODUCT = N_PRODUCT * N

        IF (INDEX_VAL < (N_MAX - 1)) NUMBERS(INDEX_VAL:N_MAX) = EOSHIFT(       &
                                     NUMBERS(INDEX_VAL:N_MAX), 1, "0")

        INDEX_VAL = MOD(FLOOR(REAL((I - 1) * N_PRODUCT)                        &
                  / REAL(MAX_PERMUTATION)), N) + 1

        DIGITS_VAL(J:J) = NUMBERS(INDEX_VAL)

    END DO

    PRINT ('(A)'), DIGITS_VAL

END PROGRAM F024
